cocos2d/Platforms/Mac/CCViewMacGL.m

```diff
@@ -0,0 +1,296 @@
+    // When resizing the view, -reshape is called automatically on the main thread
+    // Add a mutex around to avoid the threads accessing the context simultaneously when resizing
+    
+    [self lockOpenGLContext];
+    
+    NSRect rect = [self convertRectToBacking:self.bounds];
+    
+    CCDirector *director = _director;
+    [director reshapeProjection: NSSizeToCGSize(rect.size) ];
+    
+    // avoid flicker
+    // Only draw if there is something to draw, otherwise it actually creates a flicker of the current glClearColor
+    if(director.runningScene){
+        [director mainLoopBody];
+    }
+    //	[self setNeedsDisplay:YES];
+    
+    [self unlockOpenGLContext];
+}
+
+
+-(void) lockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    [glContext makeCurrentContext];
+    CGLLockContext([glContext CGLContextObj]);
+}
+
+-(void) unlockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    CGLUnlockContext([glContext CGLContextObj]);
+}
+
+// Find or make a fence that is ready to use.
+-(CCGLFence *)getReadyFence
+{
+    // First checkf oldest (first in the array) fence is ready again.
+    CCGLFence *fence = _fences.firstObject;;
+    if(fence.isReady){
+        // Remove the fence so it can be inserted at the end of the queue again.
+        [_fences removeObjectAtIndex:0];
+        return fence;
+    } else {
+        // No existing fences ready. Make a new one.
+        return [[CCGLFence alloc] init];
+    }
+}
+
+-(void)addFrameCompletionHandler:(dispatch_block_t)handler
+{
+    if(_fences == nil){
+        _fences = [NSMutableArray arrayWithObject:[[CCGLFence alloc] init]];
+    }
+    
+    CCGLFence *fence = _fences.lastObject;
+    if(!fence.isReady){
+        fence = [self getReadyFence];
+        [_fences addObject:fence];
+    }
+    
+    [fence.handlers addObject:handler];
+}
+
+-(void)beginFrame
+{
+    [self lockOpenGLContext];
+}
+
+-(void)presentFrame
+{
+    {
+        CCGLFence *fence = _fences.lastObject;
+        if(fence.isReady){
+            // If the fence is ready to be added, insert a sync point for it.
+            [fence insertFence];
+        }
+    }
+    
+    [self.openGLContext flushBuffer];
+    
+    // Check the fences for completion.
+    for(CCGLFence *fence in _fences){
+        if(fence.isCompleted){
+            for(dispatch_block_t handler in fence.handlers) handler();
+            [fence.handlers removeAllObjects];
+        } else {
+            break;
+        }
+    }
+    
+    [self unlockOpenGLContext];
+}
+
+-(GLuint)fbo
+{
+    return 0;
+}
+
+- (void) dealloc
+{
+    CCLOGINFO(@"cocos2d: deallocing %@", self);
+}
+
+#pragma mark CCGLView - Mouse Delegate
+
+- (void)mouseDown:(NSEvent *)theEvent
+{
+    // dispatch mouse to responder manager
+    [CCDirector bindDirector:_director];
```

Comment:
Yeah, unfortunately the bindDirector calls are pretty verbose in this section. The reason for this is that each one of these events is an entry point into cocos2d code from various NS input handling places. Further down this stack trace, we will have user code responding to mouseDown events. Inside that user code, the user might be looking for the `[CCDirector currentDirector]` for that thread, and it needs to be bound for them. 

I'm open to other suggestions.

It could be moved into the scheduler, but I think that's just pushing the binds into bigger methods where it's not as easy to see what's going on. (although looking at it, there would be a few less calls).

And it's important to unbind the director to ensure correctness. It makes it much easier to find a mistake like a missing bind in the engine, rather than leaving the 'usually correct' director bound, causing infrequent race-condition like issues when the user's setting up a new scene or something like that.

Thank you for reviewing!